tests/lib/config/flat-config-helpers.js

```diff
@@ -99,4 +102,125 @@ describe("Config Helpers", () => {
         });
     });
 
+    describe("getRuleOptionsSchema", () => {
+        it("should return null when no schema specified", () => {
+            /* eslint-disable no-undefined -- manually testing that missing properties and undefined both are accounted for */
+            assert.isNull(getRuleOptionsSchema({ }));
+            assert.isNull(getRuleOptionsSchema({ schema: undefined }));
+            assert.isNull(getRuleOptionsSchema({ schema: null }));
+            assert.isNull(getRuleOptionsSchema({ meta: { } }));
+            assert.isNull(getRuleOptionsSchema({ meta: { } }));
+            assert.isNull(getRuleOptionsSchema({ schema: undefined, meta: { } }));
+            assert.isNull(getRuleOptionsSchema({ schema: null, meta: { } }));
+            assert.isNull(getRuleOptionsSchema({ meta: { schema: undefined } }));
+            assert.isNull(getRuleOptionsSchema({ meta: { schema: null } }));
+            assert.isNull(getRuleOptionsSchema({ schema: undefined, meta: { schema: undefined } }));
+            assert.isNull(getRuleOptionsSchema({ schema: undefined, meta: { schema: null } }));
+            assert.isNull(getRuleOptionsSchema({ schema: null, meta: { schema: undefined } }));
+            assert.isNull(getRuleOptionsSchema({ schema: null, meta: { schema: null } }));
+            assert.isNull(getRuleOptionsSchema({ schema: undefined, meta: { } }));
+            assert.isNull(getRuleOptionsSchema({ schema: null, meta: { } }));
+            /* eslint-enable no-undefined -- manually testing that missing properties and undefined both are accounted for  */
+        });
+
+        it("should create schema from rule.schema", () => {
+            const outputSchema = getRuleOptionsSchema({ schema: [{ enum: [0] }] });
+
+            assert.exists(outputSchema);
+        });
+
+        it("should create schema from rule.meta.schema", () => {
+            const outputSchema = getRuleOptionsSchema({ meta: { schema: [{ enum: [0] }] } });
+
+            assert.exists(outputSchema);
+        });
+
+        it("should prefer rule.schema over rule.meta.schema", () => {
+            const outputSchema = getRuleOptionsSchema({
+                schema: [{ enum: [0] }],
+                meta: { schema: [{ enum: [1] }] }
+            });
+
+            assert.isTrue(ajv.validate(outputSchema, [0]));
+            assert.isFalse(ajv.validate(outputSchema, [1]));
```

Comment:
I tried to write the tests in such a way that it's testing the overall contract of the method, i.e. "given this rule, return _some_ JSON schema (or null) that will validate its sliced options array in the documented way". We don't actually care about the specific shape of the schema returned, we care about what behaviour that schema exhibits.

Sending the output schema through ajv and validating that it accepts/rejects values that we know will be valid/invalid allows the exact shape of the schema to be modified without changing the test. The test will still be testing "does the method give me a schema that will validate the options in the way we expect", but it allows flexibility of implementation.

The `$ref` PR is an example of this (ignoring whether that change actually happens) - the structure of the JSON Schema is being changed, but the test is still testing the actual behaviour of whatever structure is generated. A smaller example might be eliminating `minItems: 0`, since it's technically redundant - the behaviour of the schema would be the same, but a test asserting that it was set would start failing.

Reply:
Tests removed from PR.